client/src/main/java/io/pravega/client/connection/impl/ConnectionPoolImpl.java

```diff
@@ -0,0 +1,207 @@
+
+        final List<Connection> connectionList = connectionMap.getOrDefault(location, new ArrayList<>());
+
+        // remove connections for which the underlying network connection is disconnected.
+        List<Connection> prunedConnectionList = connectionList.stream().filter(connection -> {
+            // Filter out Connection objects which have been completed exceptionally or have been disconnected.
+            return !connection.getFlowHandler().isDone() || connection.isConnected();
+        }).collect(Collectors.toList());
+        log.debug("List of connections to {} that can be used: {}", location, prunedConnectionList);
+
+        // Choose the connection with the least number of flows.
+        Optional<Connection> suggestedConnection = prunedConnectionList.stream().min(Comparator.naturalOrder());
+
+        final Connection connection;
+        if (suggestedConnection.isPresent() && (prunedConnectionList.size() >= clientConfig.getMaxConnectionsPerSegmentStore() || isUnused(suggestedConnection.get()))) {
+            log.info("Reusing connection: {}", suggestedConnection.get());
+            connection = suggestedConnection.get();
+        } else {
+            // create a new connection.
+            log.info("Creating a new connection to {}", location);
+            CompletableFuture<FlowHandler> establishedFuture = establishConnection(location);
+            connection = new Connection(location, establishedFuture);
+            prunedConnectionList.add(connection);
+        }
+        connectionMap.put(location, prunedConnectionList);
+        return connection.getFlowHandler().thenApply(flowHandler -> flowHandler.createFlow(flow, rp));
+    }
+
+    @Override
+    @Synchronized
+    public CompletableFuture<ClientConnection> getClientConnection(PravegaNodeUri location, ReplyProcessor rp) {
+        Preconditions.checkNotNull(location, "Location");
+        Preconditions.checkNotNull(rp, "ReplyProcessor");
+        Exceptions.checkNotClosed(closed.get(), this);
+
+        // create a new connection.
+        CompletableFuture<FlowHandler> handler = establishConnection(location);
+        Connection connection = new Connection(location, handler);
+        return connection.getFlowHandler().thenApply(h -> h.createConnectionWithFlowDisabled(rp));
+    }
+
+    private boolean isUnused(Connection connection) {
+        return Futures.isSuccessful(connection.getFlowHandler()) && connection.getFlowCount() == 0;
+    }
+
+    /**
+     * Used only for testing.
+     */
+    @VisibleForTesting
+    @Synchronized
+    public void pruneUnusedConnections() {
+        for (List<Connection> connections : connectionMap.values()) {
+            for (Iterator<Connection> iterator = connections.iterator(); iterator.hasNext();) {
+                Connection connection = iterator.next();
+                if (isUnused(connection)) {
+                    connection.getFlowHandler().join().close();
```

Comment:
should we log a line ?
this .join() may take time to complete, if there is something stuck.

Reply:
If the `isUnused` function first checks `Futures.isSuccessful` so if it is considered unused the `join` should be instant.